AdminCore.WebApp/Controllers/EventController.cs

```diff
@@ -185,49 +193,67 @@ public IActionResult UpdateEvent(UpdateEventViewModel updateEventViewModel)
-        return ApproveIfEventDoesNotBelongToTheAdmin(approveEventViewModel, eventToApprove);
-      }
-      catch (Exception ex)
-      {
-        Logger.LogError(ex.Message);
-        return StatusCode((int)HttpStatusCode.InternalServerError, "Something went wrong approving event");
-      }
+      return ProcessEvent(approveEventViewModel.EventId, _eventWorkflowService.WorkflowResponseApprove);
     }
-
+    
     [HttpPut("cancelEvent")]
     public IActionResult CancelEvent(CancelEventViewModel cancelEventViewModel)
     {
-      try
-      {
-        _eventService.UpdateEventStatus(cancelEventViewModel.EventId, EventStatuses.Cancelled);
-        return Ok("Successfully Cancelled");
-      }
-      catch (Exception ex)
-      {
-        Logger.LogError(ex.Message);
-        return StatusCode((int)HttpStatusCode.InternalServerError, "Something went wrong cancelling event");
-      }
+      return ProcessEvent(cancelEventViewModel.EventId, _eventWorkflowService.WorkflowResponseCancel);
     }
 
     [Authorize("Admin")]
     [HttpPut("rejectEvent")]
     public IActionResult RejectEvent(RejectEventViewModel rejectEventViewModel)
+    {
+      return ProcessEvent(rejectEventViewModel.EventId, _eventWorkflowService.WorkflowResponseReject, rejectEventViewModel.Message);
+    }
+
+    private IActionResult ProcessEvent(int eventId, Func<EventDto, EmployeeDto, WorkflowFsmStateInfo> workflowProcessFunc, string eventMessage = null)
     {
       try
-      {
-        _eventService.RejectEvent(rejectEventViewModel.EventId, rejectEventViewModel.Message, _employee.EmployeeId);
-        return Ok("Successfully Rejected");
+      {       
+        var leaveEvent = _eventService.GetEvent(eventId);
+
+        try
+        {
+          if (leaveEvent.EventStatusId != (int) EventStatuses.AwaitingApproval)
+          {
+            return StatusCode((int) HttpStatusCode.OK, "Event is not awaiting any approval response.");
+          }
+
+          // Advance workflow.
+          var workflowResultState = workflowProcessFunc(leaveEvent, _employee);
+          // Add message to event.
+          _eventService.AddRejectMessageToEvent(eventId, eventMessage, _employee.EmployeeId);
+              
+          UpdateEventStatus(workflowResultState, eventId);
+
+          return Ok("Leave response sent successfully.\n" +
+                    $"Current event state: {workflowResultState.CurrentEventStatuses}\n" +
+                    $"Event workflow message: {workflowResultState.Message}");
+
+        }
+        catch (ValidationException e)
+        {
+          return StatusCode((int)HttpStatusCode.Forbidden, e.Message);
+        }
       }
       catch (Exception ex)
       {
         Logger.LogError(ex.Message);
-        return StatusCode((int)HttpStatusCode.InternalServerError, "Something went wrong rejecting event");
+        return StatusCode((int)HttpStatusCode.InternalServerError, "Something went wrong sending leave response for event.");
```

Comment:
AFAIK period at end of exception message is redundant

Reply:
I personally prefer to make error messages grammatically correct. I'm not sure what the best practice is.

Reply:
OK. As long as we are consistent, at the moment we are not so lets stick with ending except msgs with a `.` 😄